zkvm/src/encoding.rs

```diff
@@ -44,26 +43,45 @@ impl<'a> Subslice<'a> {
         self.start..self.end
     }
 
-    /// Copies the contents of the subslice into a `Vec<u8>`.
-    pub fn to_vec(&self) -> Vec<u8> {
-        let mut buf = Vec::with_capacity(self.len());
-        buf.extend_from_slice(&self);
-        buf
+    pub fn slice<F, T>(&mut self, slice_fn: F) -> Result<(T, &[u8]), VMError>
+    where
+        F: FnOnce(&mut SliceReader) -> Result<T, VMError>,
+    {
+        let start = self.start;
+        let result = slice_fn(self)?;
+        let end = self.start;
+        Ok((result, &self.whole[start..end]))
     }
 
-    /// Returns a Subslice of the first `prefix_size` of bytes and advances
+    pub fn parse<F, T>(data: &[u8], parse_fn: F) -> Result<T, VMError>
+    where
+        F: FnOnce(SliceReader) -> Result<T, VMError>,
```

Comment:
If we pass a &mut instead of move the reader in, probably we can also make this line cleaner:

before:
```rust
SliceReader::parse(&data, |mut r| Self::decode(&mut r))?
```
after:
```rust
SliceReader::parse(&data, |r| Self::decode(r))?
```